src/main/assisted-installer-controller/assisted_installer_main.go

```diff
@@ -55,38 +62,77 @@ func main() {
 		kc,
 	)
 
+	assistedController.SetReadyState()
+
 	// While adding new routine don't miss to add wg.add(1)
 	// without adding it will panic
 	var wg sync.WaitGroup
-	var wgLogs sync.WaitGroup
 	var status assistedinstallercontroller.ControllerStatus
 
-	ctxApprove, cancelApprove := context.WithCancel(context.Background())
-	go assistedController.ApproveCsrs(ctxApprove, &wg)
+	ctxRoutines, cancelRoutines := context.WithCancel(context.Background())
+	go assistedController.WaitAndUpdateNodesStatus(ctxRoutines, &wg)
 	wg.Add(1)
-	go assistedController.PostInstallConfigs(&wg, &status)
+	go assistedController.PostInstallConfigs(ctxRoutines, &wg, &status)
 	wg.Add(1)
-	go assistedController.UpdateBMHs(&wg)
+	go assistedController.UpdateBMHs(ctxRoutines, &wg)
 	wg.Add(1)
+
+	// No need to cancel with context, will finish quickly
 	go assistedController.HackDNSAddressConflict(&wg)
 	wg.Add(1)
 
-	ctxLogs, cancelLogs := context.WithCancel(context.Background())
-	go assistedController.UploadLogs(ctxLogs, cancelLogs, &wgLogs, &status)
-	wgLogs.Add(1)
+	go assistedController.UploadLogs(ctxRoutines, &wg, &status)
+	wg.Add(1)
 
-	assistedController.SetReadyState()
-	assistedController.WaitAndUpdateNodesStatus(&status)
-	logger.Infof("Sleeping for 10 minutes to give a chance to approve all csrs")
-	time.Sleep(10 * time.Minute)
-	cancelApprove()
+	waitForInstallation(client, logger, &status)
+	// stop all go routines
+	cancelRoutines()
 
 	logger.Infof("Waiting for all go routines to finish")
 	wg.Wait()
-	if !status.HasError() {
-		//with error the logs are canceled within UploadLogs
-		logger.Infof("closing logs...")
-		cancelLogs()
+	logger.Infof("Finished all")
+}
+
+func waitForInstallation(client inventory_client.InventoryClient, log logrus.FieldLogger, status *assistedinstallercontroller.ControllerStatus) {
+	log.Infof("monitor cluster installation status")
+	reqCtx := utils.GenerateRequestContext()
+	errCounter := 0
+	for {
+		time.Sleep(assistedinstallercontroller.GeneralWaitInterval)
+		cluster, err := client.GetCluster(reqCtx)
+		if err != nil {
+			// In case cluster was deleted or controller is not authorised
+			// we should exit controller
+			switch err.(type) {
+			case *installer.GetClusterNotFound:
```

Comment:
I would rather not Exit here but gracefully terminate the context.However, if you remain with Exit at least put here a warning not to put here other error codes. This could result in abnormal termination of the controller in spotty network conditions with out completing the logs

Reply:
Not sure you understood the issue. There is no cluster in the system, there is nothing to update. No api calls will work. The idea is to just exit.

Reply:
This is the current issue. My concern is that in 3 months time someone will come and add conditions here that are not kubeapi related.

Reply:
I think that if you can't run apis , there is no need to try. What you are asking will require adding logic to routines to handle this specific case and that what i want to avoid